### ISSUE TITLE
OSError: [Errno 30] Read-only file system: '/boot/efi/nixos/vgbwi0dnqv79j3k051myc6vbyqkjwns6-linux-4.9.37-bzImage.efi'

### ISSUE THREAD

**Issue by ixxie** [NixOS/nixpkgs#28817]
After messing around with the wrapper scripts I use for rebuilding my system, suddenly most of my generations no longer boot, and rebuilds (this time the normal kind without a script) no longer work (for a configuration that built in the past). They yield the following error:

    meso% sudo nixos-rebuild switch
    building Nix...
    building the system configuration...
    Traceback (most recent call last):
      File "/nix/store/3qbr1aa4psdiwhsfmmsk6pab82l0a6d5-systemd-boot-builder.py", line 149, in <module>
    main()
      File "/nix/store/3qbr1aa4psdiwhsfmmsk6pab82l0a6d5-systemd-boot-builder.py", line 142, in main
    remove_old_entries(gens)
      File "/nix/store/3qbr1aa4psdiwhsfmmsk6pab82l0a6d5-systemd-boot-builder.py", line 105, in remove_old_entries
        os.unlink(path)
    OSError: [Errno 30] Read-only file system: '/boot/efi/nixos/vgbwi0dnqv79j3k051myc6vbyqkjwns6-linux-4.9.37-bzImage.efi'
    warning: error(s) occurred while switching to the new configuration

The last line - `OSError: [Errno 30] Read-only file system: '/boot/efi/nixos/vgbwi0dnqv79j3k051myc6vbyqkjwns6-linux-4.9.37-bzImage.efi'` is the error I get when trying to boot most of my generations.

This leads me to believe I have done something *monumentally* stupid with my scripts. For completeness, they look something like this (`update-rebuild` is the one I ran just before the trouble)

`update-rebuild`:  `sudo -i nix-channel --update && rebuild-system && clean-system && exit`

`rebuild-system`:  `sudo cp -R /my/config/directory/* /etc/nixos && sudo nixos-rebuild switch`

`clean-system`:    `nix-collect-garbage --delete-older-than 31d`

I used to have the scripts have no `sudo`s in them, but I wanted to install them declaratively and tried this approach.

* System: NixOS 17.03.1517.849b5a5193 (Gorilla)
* Nix version: 1.11.11
* Nixpkgs version: "17.03.1775.56da88a298"
* Sandboxing enabled: nope

**Comment by ixxie** [comment]
My `hardware-configuration.nix`:

    # Do not modify this file!  It was generated by ‘nixos-generate-config’
    # and may be overwritten by future invocations.  Please make changes
    # to /etc/nixos/configuration.nix instead.
    { config, lib, pkgs, ... }:

    {
      imports =
        [ <nixpkgs/nixos/modules/installer/scan/not-detected.nix>
        ];

      boot.initrd.availableKernelModules = [ "xhci_pci" "ahci" "nvme" "usb_storage" "sd_mod" "rtsx_pci_sdmmc" ];
      boot.kernelModules = [ "kvm-intel" ];
      boot.extraModulePackages = [ ];

      fileSystems."/" =
        { device = "/dev/disk/by-uuid/5dc6254c-9c02-4695-8f70-7b9d19f929e7";
          fsType = "ext4";
        };

      fileSystems."/boot" =
        { device = "/dev/disk/by-uuid/9E00-97EC";
          fsType = "vfat";
        };

      swapDevices = [ ];

      nix.maxJobs = lib.mkDefault 4;
    }

**Comment by ixxie** [comment]
Alright, @grahamc helped me diagnose and resolve the issue. The following is what we did:

**1. Check `/boot` is mounted**

We checked that `/boot` was mounted using `mount`; in the output, the line: 

    "/dev/nvme0n1p1 on /boot type vfat (ro,relatime,fmask=0022,dmask=0022,codepage=437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro)"

**2. Looked for errors relating to '/boot'**

Used `dmesg | grep nvme0n1p1` to look for errors related to `/boot` in the kernel message buffer; this yielded:

    [    2.732212] FAT-fs (nvme0n1p1): Volume was not properly unmounted. Some data may be corrupt. Please run fsck.
    [ 1535.306323] FAT-fs (nvme0n1p1): error, fat_free_clusters: deleting FAT entry beyond EOF
    [ 1535.306325] FAT-fs (nvme0n1p1): Filesystem has been set read-

**3. Unmounted '/boot' and ran `fsck` on it**

Unmounted `/boot` and checked for filesystem errors using the following:

    sudo umount /boot
    sudo fsck nvme0n1p1

The check found a dirty bit and a mismatch between some sector counts, and I followed the recommended steps to correct it.

**4. Remounted '/boot' and rebooted the system**

Remounting:

    sudo mount /dev/nvme0n1p1 /boot

and voila! Problem solved. Rebooted and now all my generations work again.
